src/main/java/gregtech/common/crafting/DyeableRecipes.java

```diff
@@ -0,0 +1,140 @@
+package gregtech.common.crafting;
+
+import gregtech.api.items.IDyeableItem;
+
+import net.minecraft.inventory.InventoryCrafting;
+import net.minecraft.item.ItemStack;
+import net.minecraft.item.crafting.RecipesArmorDyes;
+import net.minecraft.util.NonNullList;
+import net.minecraft.world.World;
+
+import it.unimi.dsi.fastutil.objects.ObjectArrayList;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.List;
+
+/**
+ * Composed almost entirely of code taken from {@link RecipesArmorDyes} and modified to apply to {@link IDyeableItem}
+ * instead of just leather armor. Thus, will be difficult to parse.
+ */
+public final class DyeableRecipes extends RecipesArmorDyes {
+
+    @ApiStatus.Internal
+    public DyeableRecipes() {
+        this.setRegistryName("gt_dyeable_dyeing");
+    }
+
+    /**
+     * Used to check if a recipe matches current crafting inventory
+     */
+    @Override
+    public boolean matches(InventoryCrafting inv, @NotNull World worldIn) {
+        ItemStack itemstack = ItemStack.EMPTY;
+        List<ItemStack> list = new ObjectArrayList<>();
+
+        for (int i = 0; i < inv.getSizeInventory(); ++i) {
+            ItemStack itemstack1 = inv.getStackInSlot(i);
+
+            if (!itemstack1.isEmpty()) {
+                if (itemstack1.getItem() instanceof IDyeableItem) {
+                    if (!itemstack.isEmpty()) {
+                        return false;
+                    }
+
+                    itemstack = itemstack1;
+                } else {
+                    if (!net.minecraftforge.oredict.DyeUtils.isDye(itemstack1)) {
+                        return false;
+                    }
+
+                    list.add(itemstack1);
+                }
+            }
+        }
+
+        return !itemstack.isEmpty() && !list.isEmpty();
+    }
+
+    /**
+     * Returns an Item that is the result of this recipe
+     */
+    @Override
+    public @NotNull ItemStack getCraftingResult(InventoryCrafting inv) {
+        ItemStack itemstack = ItemStack.EMPTY;
+        int[] aint = new int[3];
+        int i = 0;
+        int j = 0;
+        IDyeableItem dyeable = null;
+
+        for (int k = 0; k < inv.getSizeInventory(); ++k) {
+            ItemStack itemstack1 = inv.getStackInSlot(k);
+
+            if (!itemstack1.isEmpty()) {
+                if (itemstack1.getItem() instanceof IDyeableItem) {
+                    dyeable = (IDyeableItem) itemstack1.getItem();
+
+                    itemstack = itemstack1.copy();
+                    itemstack.setCount(1);
+
+                    if (dyeable.hasColor(itemstack1)) {
+                        int l = dyeable.getColor(itemstack);
```

Comment:
We do have our own DyeUtil#determineDyeColor, if that would help cleanup this mess